gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/avro/AvroKeyDedupReducer.java

```diff
@@ -28,31 +35,50 @@
  *
  * @author Ziyang Liu
  */
-public class AvroKeyDedupReducer
-    extends Reducer<AvroKey<GenericRecord>, AvroValue<GenericRecord>, AvroKey<GenericRecord>, NullWritable> {
+public class AvroKeyDedupReducer extends Reducer<AvroKey<GenericRecord>, AvroValue<GenericRecord>, AvroKey<GenericRecord>, NullWritable> {
 
   public enum EVENT_COUNTER {
     MORE_THAN_1,
     DEDUPED,
     RECORD_COUNT
   }
 
+  public static final String DELTA_SCHEMA_PROVIDER =
+      "gobblin.compaction." + AvroKeyDedupReducer.class.getSimpleName() + ".deltaFieldsProvider";
   private AvroKey<GenericRecord> outKey;
+  private Optional<AvroValueDeltaSchemaComparator> deltaComparatorOptional;
+  private AvroDeltaFieldNameProvider deltaFieldNamesProvider;
 
   @Override
-  protected void setup(Context context) throws IOException, InterruptedException {
+  protected void setup(Context context)
+      throws IOException, InterruptedException {
     this.outKey = new AvroKey<>();
+    this.deltaComparatorOptional = Optional.absent();
+    Configuration conf = context.getConfiguration();
+    String deltaSchemaProviderClassName = conf.get(DELTA_SCHEMA_PROVIDER);
+    if (deltaSchemaProviderClassName != null) {
+      this.deltaFieldNamesProvider =
+          GobblinConstructorUtils.invokeConstructor(AvroDeltaFieldNameProvider.class, deltaSchemaProviderClassName, conf);
+      this.deltaComparatorOptional = Optional.of(new AvroValueDeltaSchemaComparator(deltaFieldNamesProvider));
+    }
   }
 
   @Override
   protected void reduce(AvroKey<GenericRecord> key, Iterable<AvroValue<GenericRecord>> values, Context context)
       throws IOException, InterruptedException {
     int numVals = 0;
 
+    AvroValue<GenericRecord> valueToRetain = null;
+
     for (AvroValue<GenericRecord> value : values) {
-      this.outKey.datum(value.datum());
+      if (valueToRetain == null) {
+        valueToRetain = value;
+      } else if (this.deltaComparatorOptional.isPresent()) {
+        valueToRetain = this.deltaComparatorOptional.get().compare(valueToRetain, value) > 0 ? valueToRetain : value;
```

Comment:
Nit: if there are multiple records that are equal, this algorithm will chose the first record if `deltaComparatorOptional` is missing, but it will chose the last record if it is present. For consistency sake, can you change the condition to `>=`?